src/TropicalGeometry/semiring.jl

```diff
@@ -0,0 +1,504 @@
+# module Tropical
+
+export tropical_semiring,
+       @tropical,
+       determinant,
+       permanent
+
+# using Reexport
+
+# @reexport using Oscar
+
+# import Oscar: expressify, Ring, @enable_all_show_via_expressify, zero, one, iszero, isone, PolynomialRing
+
+################################################################################
+#
+#  Types
+#
+################################################################################
+
+# We use T to record whether we are in the min/max case
+# T is either typeof(min) or typeof(max)
+mutable struct TropicalSemiring{T} <: Field
```

Comment:
Is there a reason this is `mutable`? Otherwise, just drop it, then it becomes a singleton type, which may enable further optimizations.
```suggestion
struct TropicalSemiring{T} <: Field
```